### DISCUSSION TITLE
gerbera 2.3.0 and config parms for scripts?

### DISCUSSION THREAD

**Discussion by nobanzai** [gerbera/gerbera#3265] [General]
Hi *,

up to gebera 2.2.0 I used to retrieve my autoscan directories in the addAudio function via the config hash.
In version 2.3.0 this method doesn't work anymore, as Karl pointed out in some posting - "no more globals".
Is there a different method to get this information in addAudio when using <import-function>?

Btw. is it the correct method to just overload addAudio in my own js-Script, i.e. let gerbera's importAudio function just call my addAudio instead of the default one?

Thx and bye.
Michael.

**Comment by KarlStraussberger** [comment]
You don't have to overwrite anymore.

You can place your code in a new file which is loaded from custom-script-folder (https://docs.gerbera.io/en/latest/config-import.html#script-folder) and set your function name as import-function `importAudioNNN` (https://docs.gerbera.io/en/latest/config-import.html#import-function)

Your code may follow the pattern:

```
var object_ref_list;
function importAudioNNN(obj, cont, rootPath, autoscanId, containerType) {
    object_autoscan_id = autoscanId;
    object_ref_list = addAudioNNN(obj, cont, rootPath, containerType);
    return object_ref_list;
}
function addAudioNNN(obj, cont, rootPath, containerType) {
  ... your code
}
```

**Reply by nobanzai** [reply]
Why do we need object_ref_list and object_autoscan_id?
Isn't
```
function importAudio_mh001(obj, cont, rootPath, autoscanId, containerType) {
    return addAudio_mh001(obj, cont, rootPath, containerType);
}  
```
enough?

**Reply by nobanzai** [reply]
I am not even sure, why the default addAudio function needs a result variable to return.

**Reply by KarlStraussberger** [reply]
The first is the result.

The second some legacy paranoia. You can drop it if you don't use it in your script.

**Reply by KarlStraussberger** [reply]
The result is the list of created objects. It is used to detect objects that are not needed anymore. Look at default scripts: `result.push(addCdsObject(obj, container));`

**Reply by nobanzai** [reply]
Yes, that is, why I asked. In my addAudio I used "addCdsObject(obj, container)" but never assigned the return value to any variable. And I also never returned anything from the function. Nevertheless the resulting database content is/was exactly the same.

**Reply by KarlStraussberger** [reply]
That changes if you upgrade to 2.3, afaik.

**Reply by nobanzai** [reply]
In the meantime I got 2.3 :)
But nothing changed.

**Reply by KarlStraussberger** [reply]
It's in the autoscan code to remove items that have not been updated by the script.

**Reply by nobanzai** [reply]
Ok, to understand that - without returning the hash, everything is built up, but deleted or moved items are not purged from the database?

**Reply by KarlStraussberger** [reply]
Everything is built up on first scan.

On update all elements that cannot be found in the result will be deleted - you'll have to restart to get them back.

**Reply by nobanzai** [reply]
May it cause "No Container parent for Item " messages, if I don't use the result hash, too?

**Reply by KarlStraussberger** [reply]
Quite possible if objects are getting un-rooted.

**Comment by nobanzai** [comment]
I didn't overwrite anything.
I placed an xy.js in an extra directory, that was added by

      <script-folder>
        <common>/usr/share/gerbera/js</common>
        <custom><mypath>/gerbera</custom>
      </script-folder>

import function wasn't modifed

      <import-function>
        <audio-file>importAudio</audio-file>

and xy.js contains an addAudio function.
Should I better use your alternative?

**Reply by KarlStraussberger** [reply]
By that you are overwriting the default function `addAudio` which relies on the loading sequence of the files.

Since the changes in config we are able to provide additional functions. Also, you can easily switch back by the config file.

**Comment by nobanzai** [comment]
But what about the config hash?

**Reply by KarlStraussberger** [reply]
Which config hash do you mean?

**Comment by nobanzai** [comment]
In this case config['/import/autoscan/directory'].

**Reply by KarlStraussberger** [reply]
You can access all config variables in that way. Should be fully available

**Reply by KarlStraussberger** [reply]
There should be the whole array of autoscan objects in `config['/import/autoscan]`

**Reply by nobanzai** [reply]
No, config['/import/autoscan'] is undefined.
At least this is, what
print2( "Info", config['/import/autoscan'] );
tells me.

**Reply by KarlStraussberger** [reply]
Please check out `config['/import/autoscan/timed/directory']` and `config['/import/autoscan/inotify/directory']` and tell me where to add this to the docs.

**Reply by nobanzai** [reply]
Both are undefined.

**Reply by KarlStraussberger** [reply]
I'll have to take a look at my extended debug log

**Reply by nobanzai** [reply]
Did you find the time to do that?

**Reply by KarlStraussberger** [reply]
I double-checked everything and could not really see an issue. For me `print2("Debug", "autoscan ", JSON.stringify(config['/import/autoscan/inotify/directory']));` produced the expected output.

In #3344 the arrays are correctly set (only the dirs with the correct scan mode).

**Comment by nobanzai** [comment]
config['/import/autoscan/directory'] is always empty.
And didn't you write, that globals are no longer available to the addAudio and similar functions?

**Reply by KarlStraussberger** [reply]
That's why the variables are created and assigned in the script.
